src/webauthn/authenticator/mod.rs

```diff
@@ -0,0 +1,506 @@
+use rsa::{
+    pkcs1::{DecodeRsaPrivateKey, EncodeRsaPrivateKey},
+    signature::SignatureEncoding,
+    traits::PublicKeyParts,
+};
+use serde_cbor::Value;
+use sha1::Sha1;
+use sha2::Sha256;
+use uuid::Uuid;
+
+use crate::webauthn::{
+    authenticator::responses::PrivateKeyResponse,
+    proto::{
+        constants::WEBAUTHN_REQUEST_TYPE_GET,
+        web_message::{get_default_rp_id, AuthenticatorAttestationResponseRaw, PublicKeyCredentialRaw, PublicKeyCredentialRequestOptions},
+    },
+};
+#[cfg(test)]
+use crate::webauthn::{
+    proto::web_message::{
+        Extensions, PublicKeyCredentialDescriptor, PublicKeyCredentialRpEntity, PublicKeyCredentialType, PublicKeyCredentialUserEntity,
+    },
+    server::{CredentialCreationVerifier, CredentialRequestVerifier},
+};
+
+#[derive(Debug)]
+pub enum WebauthnCredentialRequestError {
+    UserVerificationRequired,
+    AlgorithmNotSupported,
+    CouldNotGenerateKey,
+    RpIdOrOriginRequired,
+    RpIdHashInvalidLength(usize),
+    SerdeJsonError(serde_json::Error),
+    SerdeCborError(serde_cbor::Error),
+    WebauthnError(Error),
+    RsaError(rsa::pkcs1::Error),
+    Base64Error(base64::DecodeError),
+    Ed25519Error(ed25519_dalek::SignatureError),
+}
+
+impl From<serde_json::Error> for WebauthnCredentialRequestError {
+    fn from(e: serde_json::Error) -> Self {
+        WebauthnCredentialRequestError::SerdeJsonError(e)
+    }
+}
+
+impl From<serde_cbor::Error> for WebauthnCredentialRequestError {
+    fn from(e: serde_cbor::Error) -> Self {
+        WebauthnCredentialRequestError::SerdeCborError(e)
+    }
+}
+
+impl From<Error> for WebauthnCredentialRequestError {
+    fn from(e: Error) -> Self {
+        WebauthnCredentialRequestError::WebauthnError(e)
+    }
+}
+
+impl From<rsa::pkcs1::Error> for WebauthnCredentialRequestError {
+    fn from(value: rsa::pkcs1::Error) -> Self {
+        WebauthnCredentialRequestError::RsaError(value)
+    }
+}
+
+impl From<base64::DecodeError> for WebauthnCredentialRequestError {
+    fn from(value: base64::DecodeError) -> Self {
+        WebauthnCredentialRequestError::Base64Error(value)
+    }
+}
+
+impl From<SignatureError> for WebauthnCredentialRequestError {
+    fn from(value: SignatureError) -> Self {
+        WebauthnCredentialRequestError::Ed25519Error(value)
+    }
+}
+
+pub struct WebauthnAuthenticator;
+
+impl WebauthnAuthenticator {
+    pub fn generate_credential_creation_response(
+        credential_creation_options: PublicKeyCredentialCreationOptions,
+        aaguid: Uuid,
+        credential_id: Vec<u8>,
+        origin: Option<String>,
+        attestation_flags: u8,
+    ) -> Result<AuthenticatorCredentialCreationResponse, WebauthnCredentialRequestError> {
+        if credential_creation_options
+            .authenticator_selection
+            .as_ref()
+            .and_then(|auth_selection| auth_selection.user_verification.as_ref())
+            .filter(|user_verif| *user_verif == &UserVerificationRequirement::Required)
```

Comment:
To make sure you're not comparing an intermediate reference instead of the data itself, I'd do as such instead of the weird hybrid of dereferencing one level of reference but putting a reference on the other side:
```suggestion
            .filter(|user_verif| **user_verif == UserVerificationRequirement::Required)
```